File: Examples/ObjectiveC/DDSLayerSelectionExample.m
//
//  DDSLayerSelectionExample.m
//  Examples
//
//  Created by Jordan Kiley on 3/21/17.
//  Copyright © 2017 Mapbox. All rights reserved.
//

// url to use: 

#import "DDSLayerSelectionExample.h"
@import Mapbox;

NSString const *MBXExampleDDSLayerSelection = @"DDSLayerSelectionExample";

@interface DDSLayerSelectionExample () <MGLMapViewDelegate, UIGestureRecognizerDelegate>

@property (nonatomic) MGLMapView *mapView;

@end

@implementation DDSLayerSelectionExample

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.mapView = [[MGLMapView alloc] initWithFrame:self.view.bounds];
    self.mapView.delegate = self;
    
    self.mapView.autoresizingMask = UIViewAutoresizingFlexibleHeight | UIViewAutoresizingFlexibleWidth;
    [self.view addSubview:self.mapView];
    
    UITapGestureRecognizer *gesture = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(handleTap:)];
    gesture.delegate = self;
    gesture.numberOfTapsRequired = 1;
    [self.mapView addGestureRecognizer:gesture];
}

- (void)handleTap:(UITapGestureRecognizer *)gesture {
    CGPoint spot = [gesture locationInView:self.mapView];
    NSArray *features = [self.mapView visibleFeaturesAtPoint:spot inStyleLayersWithIdentifiers:[NSSet setWithObject:@"state-layer"]];
    
    MGLPolygonFeature *feature = [features firstObject];
    
    NSString *state = [feature attributeForKey:@"name"];
    [self changeOpacityBasedOn:state];
}


- (void)changeOpacityBasedOn:(NSString*)name {
    MGLFillStyleLayer *layer = [self.mapView.style layerWithIdentifier:@"state-layer"];
    if ([name length] > 0) {
        layer.fillOpacity = [MGLStyleValue valueWithInterpolationMode:MGLInterpolationModeCategorical sourceStops:@{name: [MGLStyleValue valueWithRawValue:@1]} attributeName:@"name" options:@{MGLStyleFunctionOptionDefaultValue: [MGLStyleValue valueWithRawValue:@0]}];
    } else {
        layer.fillOpacity = [MGLStyleValue valueWithRawValue:@1];
    }
}

- (void)mapView:(MGLMapView *)mapView didFinishLoadingStyle:(MGLStyle *)style {
    
    NSURL *url = [NSURL URLWithString:@"mapbox://examples.69ytlgls"];
    
    MGLVectorSource *source = [[MGLVectorSource alloc] initWithIdentifier:@"state-source" configurationURL:url];
    [style addSource:source];
    
    MGLFillStyleLayer *layer = [[MGLFillStyleLayer alloc] initWithIdentifier:@"state-layer" source:source];
    layer.sourceLayerIdentifier = @"stateData_2-dx853g";
    
    NSDictionary *stops = @{
                            @0: [MGLStyleValue valueWithRawValue:[UIColor colorWithRed:0.94 green:0.93 blue:0.96 alpha:1.0]],
                            @600: [MGLStyleValue valueWithRawValue:[UIColor colorWithRed:0.62 green:0.60 blue:0.78 alpha:1.0]],
                            @1200: [MGLStyleValue valueWithRawValue:[UIColor colorWithRed:0.33 green:0.15 blue:0.56 alpha:1.0]]
                            };
    layer.fillColor = [MGLStyleValue valueWithInterpolationMode:MGLInterpolationModeExponential sourceStops:stops attributeName:@"density" options:@{MGLStyleFunctionOptionDefaultValue : [MGLStyleValue valueWithRawValue:[UIColor whiteColor]]}];
    
    MGLStyleLayer *symbolLayer = [style layerWithIdentifier:@"place-city-sm"];
    
    [style insertLayer:layer belowLayer:symbolLayer];
}

@end

File: Examples/Swift/DDSLayerSelectionExample.swift
//
//  DDSLayerSelectionExample.swift
//  Examples
//
//  Created by Jordan Kiley on 3/21/17.
//  Copyright © 2017 Mapbox. All rights reserved.
//

import Mapbox

@objc(DDSLayerSelectionExample_Swift)

class DDSLayerSelectionExample_Swift: UIViewController, MGLMapViewDelegate, UIGestureRecognizerDelegate {
    
    var mapView : MGLMapView!
    var isStateSelected : Bool = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        mapView = MGLMapView(frame: view.bounds)
        mapView.delegate = self
        mapView.autoresizingMask = [.flexibleHeight, .flexibleWidth]
        view.addSubview(mapView)
        
        let gesture = UITapGestureRecognizer(target: self, action: #selector(handleTap(_:)))
        gesture.delegate = self
        mapView.addGestureRecognizer(gesture)
    }
    
    func handleTap(_ gesture: UITapGestureRecognizer) {
        let spot = gesture.location(in: mapView)
        let features = mapView.visibleFeatures(at: spot, styleLayerIdentifiers: Set(["state-layer"]))
        
        if let feature = features.first, let state = feature.attribute(forKey: "name") as? String{
            changeOpacity(name: state, finished: {
                isStateSelected = !isStateSelected
            })
        }
    }
    
    func changeOpacity(name: String, finished: ()->()) {
        
        let layer = mapView.style?.layer(withIdentifier: "state-layer") as! MGLFillStyleLayer
        if !isStateSelected  && name.characters.count > 0 {
                layer.fillOpacity = MGLStyleValue(interpolationMode: .categorical, sourceStops: [name : MGLStyleValue<NSNumber>(rawValue: 1)], attributeName: "name", options: [.defaultValue : MGLStyleValue<NSNumber>(rawValue: 0)])
        
        } else {
            layer.fillOpacity = MGLStyleValue(rawValue: 1)
        }
        finished()
    }
    
    func mapView(_ mapView: MGLMapView, didFinishLoading style: MGLStyle) {
        
        let url = URL(string: "mapbox://examples.69ytlgls")!
        let source = MGLVectorSource(identifier: "state-source", configurationURL: url)
        style.addSource(source)
        
        let layer = MGLFillStyleLayer(identifier: "state-layer", source: source)
        layer.sourceLayerIdentifier = "stateData_2-dx853g"
        let stops = [0: MGLStyleValue<UIColor>(rawValue: .yellow),
                     100: MGLStyleValue<UIColor>(rawValue: .red),
                     1200: MGLStyleValue<UIColor>(rawValue: .blue)]
        
        layer.fillColor = MGLStyleValue(interpolationMode: .exponential, sourceStops: stops, attributeName: "density", options: [.defaultValue : MGLStyleValue<UIColor>(rawValue: .white)])
        let symbolLayer = style.layer(withIdentifier: "state-label-sm")
        style.insertLayer(layer, below: symbolLayer!)
    }
}
